SI-6288 Perfecting positions

retronym opened the pull request:
Improves the debugging experience by fixing positions of extractor calls in patterns, and of label jumps emitted by patterns.

Review by @adriaanm (first commit), and @dragos (second)

retronym committed:
SI-6288 Position argument of unapply

`atPos(pos) { ... }` doesn't descend into children of
already positioned trees, we need to manually set the
position of `CODE.REF(binder)` to that of the stunt double
`Ident(nme.SELECTOR_DUMMY)`.

retronym committed:
SI-6288 Fix positioning of label jumps

ICode generation was assigning the position of
the last label jump to all jumps to that particular
label def.

This problem is particularly annoying under the new
pattern matcher: a breakpoint in the body of the final
case will be triggered on the way out of the body of
any other case.

Thanks to @dragos for the expert guidance as we
wended our way through GenICode to the troublesome
code. Chalk up another bug for mutability.

I believe that the ICode output should be stable
enough to use a a .check file, if it proves otherwise
we should make it so.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1142/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1142/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1856/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1856/

retronym committed:
SI-6288 Remedy ill-positioned extractor binding.

The call to `Option#get` on the result of the
unapply method was unpositioned and ended up
with the position of the `match`.

retronym committed:
Expand pattern match position tests.

 - Adds tests for unapplySeq and unapply: Boolean.
   Both seem to be well positioned after the previous
   changes.

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1861/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1861/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1146/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1146/

adriaanm commented:
PLS REBUILD ALL

adriaanm commented:
LGTM

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1874/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1874/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1159/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1159/

dragos commented:
LGTM! Very glad this kind of bugs get more attention, often a minor thing like this can be a huge time waster.
